File: built-in plugins dsp/mb-mic-sm58.cpp
/**
 * MB SM58 Modeler
 * Category : effect
 * Type     : microphone
 * Version  : 1.0.0
 * Author   : Max Booster
 * Desc     : Shure SM58 stage mic
 *
 * Part of Max Booster Built-In Plugins DSP
 */

#ifndef MB_MIC_SM58_H
#define MB_MIC_SM58_H

#include <algorithm>
#include <cmath>
#include <cstring>
#include "PluginBase.h"

class MbMicSm58 : public PluginBase {
public:
    static constexpr const char* PLUGIN_ID      = "mb-mic-sm58";
    static constexpr const char* PLUGIN_NAME    = "MB SM58 Modeler";
    static constexpr const char* PLUGIN_TYPE    = "microphone";
    static constexpr const char* PLUGIN_CATEGORY = "effect";
    static constexpr const char* VERSION         = "1.0.0";

    struct Parameters {
    float presence = 0.5f;  // range [0, 1]
    float proximity = 0.4f;  // range [0, 1]
    };

    MbMicSm58() = default;
    ~MbMicSm58() override = default;

    void setSampleRate(double sampleRate) override {
        sampleRate_ = sampleRate;
        reset();
    }

    void reset() override {
        std::memset(buffer_, 0, sizeof(buffer_));
    }

    void process(float* left, float* right, int numSamples, Parameters params) {
        params.presence = std::clamp(params.presence, 0f, 1f);
        params.proximity = std::clamp(params.proximity, 0f, 1f);
        for (int i = 0; i < numSamples; ++i) {
            left[i]  = processSample(left[i],  params);
            right[i] = processSample(right[i], params);
        }
    }

private:
    double sampleRate_ = 44100.0;
    float  buffer_[65536] = {};

    inline float processSample(float input, const Parameters& params) {
        // DSP implementation for MB SM58 Modeler
        return input;
    }
};

#endif // MB_MIC_SM58_H

File: built-in plugins dsp/mb-mic-sm58.ts
import type { PluginDefinition } from '../server/services/pluginHostService';

const MbMicSm58Plugin: PluginDefinition = { id: 'mb-mic-sm58', slug: 'mb-mic-sm58', name: 'MB SM58 Modeler', category: 'effect', type: 'microphone', version: '1.0.0', description: 'Shure SM58 stage mic', author: 'Max Booster', grade: 'A', parameters: [{ id: 'presence', name: 'Presence', type: 'float', defaultValue: 0.5, minValue: 0, maxValue: 1, automatable: true }, { id: 'proximity', name: 'Proximity', type: 'float', defaultValue: 0.4, minValue: 0, maxValue: 1, automatable: true }], defaultPreset: { presence: 0.5, proximity: 0.4 } };

export default MbMicSm58Plugin;
